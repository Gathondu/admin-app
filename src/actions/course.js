import * as CourseActionTypes from "../actionTypes/course";
import courseApi from "../api/mockCourse";
import { ajaxCallError, beginAjaxCall } from "./ajaxStatus";
import toastr from "toastr";

export const createCourseSuccess = course => {
  return { type: CourseActionTypes.CREATE_COURSE_SUCCESS, course };
};

export const deleteCourseSuccess = id => {
  return { type: CourseActionTypes.DELETE_COURSE_SUCCESS, id };
};

export const loadCoursesSuccess = courses => {
  return { type: CourseActionTypes.LOAD_COURSES_SUCCESS, courses };
};

export const updateCourseSuccess = course => {
  return { type: CourseActionTypes.UPDATE_COURSE_SUCCESS, course };
};

export const getNumberOfCoursesSuccess = total => {
  return { type: CourseActionTypes.GET_NUMBER_OF_COURSES_SUCCESS, total };
};

export const updateTotalCourses = action => {
  return { type: CourseActionTypes.UPDATE_NUMBER_OF_COURSES_SUCCESS, action };
};

export function loadCourses() {
  return function(dispatch) {
    dispatch(beginAjaxCall());
    return courseApi
      .getAllCourses()
      .then(courses => {
        dispatch(loadCoursesSuccess(courses));
        dispatch(getNumberOfCoursesSuccess(courses.length));
      });
  };
}

export function saveCourse(course) {
  return function(dispatch) {
    dispatch(beginAjaxCall());
    return courseApi
      .saveCourse(course)
      .then(savedCourse => {
        if (course.id) {
          dispatch(updateCourseSuccess(savedCourse));
        } else {
          dispatch(createCourseSuccess(savedCourse));
          dispatch(updateTotalCourses("add"));
        }
      })
      .catch(error => {
        dispatch(ajaxCallError(error));
        throw error;
      });
  };
}

export function deleteCourse(id) {
  return function(dispatch) {
    dispatch(deleteCourseSuccess(id));
    return courseApi
      .deleteCourse(id)
      .then(
        dispatch(updateTotalCourses("delete")),
        toastr.success("Course Deleted.")
      );
  };
}
