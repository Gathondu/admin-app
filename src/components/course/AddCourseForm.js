import React from 'react';
import PropTypes from 'prop-types';

let courseTitle = React.createRef();
let courseDescription = React.createRef();
const AddCourseForm = ({addCourse}) => {
    let handleSubmit = (e) => {
        e.preventDefault();
        addCourse(courseTitle.current.value, courseDescription.current.value);
        e.currentTarget.reset();
    };

    return (
        <>
        <h2>Add Course</h2>
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder="Course Title"
                ref={courseTitle} />
            <input
                type="text"
                placeholder="Course Description"
                ref={courseDescription} />
            <input
                type="submit"
                value="Save"/>
        </form>
        </>
    );
};

AddCourseForm.propTypes = {
    addCourse: PropTypes.func.isRequired
};

export default AddCourseForm;
