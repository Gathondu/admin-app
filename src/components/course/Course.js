import React from 'react';
import PropTypes from 'prop-types';

const Course = ({title, description, index, removeCourse}) => {
    return (
        <div className="course">
            <span className="course-title">
                <button className="remove-course" onClick={() => removeCourse(index)}>Ã—</button>
                {title}
                <br/>
                {description}
            </span>
        </div>
    );
};

Course.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    removeCourse: PropTypes.func.isRequired
};

export default Course;
