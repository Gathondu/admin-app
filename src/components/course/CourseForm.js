import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../common/TextInput'
import SelectInput from '../common/SelectInput'

const CourseForm = ({course, allAuthors, onSave, onChange, loading, errors, formIsValid}) => {
    return (
        <form>
            <h1>Manage Course</h1>
            <TextInput
                label="Title"
                onChange={onChange}
                name="title"
                value={course.title}
                error={errors.title}
                placeholder="Enter Course Title"/>
            <SelectInput
                label="Author"
                onChange={onChange}
                name="authorId"
                value={course.authorId}
                defaultOption="Select Author"
                options={allAuthors}
                error={errors.authorId}/>
            <TextInput
                label="Category"
                onChange={onChange}
                name="category"
                value={course.category}
                error={errors.category}
                placeholder="Enter Course Category"/>
            <TextInput
                label="Length"
                onChange={onChange}
                name="length"
                value={course.length}
                error={errors.length}
                placeholder="Enter Course Length"/>
            <input
                type="submit"
                disabled={loading || !formIsValid}
                value={loading ? 'Saving...' : 'Save'}
                className="btn btn-primary"
                onClick={onSave}/>
        </form>
    );
};

CourseForm.propTypes = {
    course: PropTypes.object.isRequired,
    allAuthors: PropTypes.array,
    onSave: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    loading: PropTypes.bool,
    errors: PropTypes.object
};
export default CourseForm;
