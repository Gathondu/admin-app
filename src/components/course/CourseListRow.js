import React from 'react';
import PropTypes from 'prop-types';
import {Link} from "react-router-dom";

const CourseListRow = ({course, deleteCourse}) => {
    return (
        <tr>
            <td><i className="fa fa-trash" onClick={() => deleteCourse(course.id)}/></td>
            <td><a href={course.watchHref} target="_blank" rel="noopener noreferrer"><i className="fa fa-television"/></a></td>
            <td><Link to={`/course/${course.id}`}>{course.title}</Link></td>
            <td>{course.authorId}</td>
            <td>{course.category}</td>
            <td>{course.length}</td>
        </tr>
    );
};

CourseListRow.propTypes = {
    course: PropTypes.object.isRequired
};

export default CourseListRow;
