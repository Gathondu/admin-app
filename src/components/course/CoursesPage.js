import React, {Component} from 'react';
import PropTypes from 'prop-types';
import AddCourseForm from "./AddCourseForm";
import Course from './Course';

class CoursesPage extends Component {
    render() {
        const {
            courses,
            addCourse,
            removeCourse
        } = this.props;
        const courseComponents = courses.map((course, index) => (
            <Course
                title={course.title}
                description={course.description}
                key={index}
                index={index}
                removeCourse={removeCourse}
            />
        ));
        return (
            [
                <h1>{this.props.title}</h1>,
                [courseComponents],
                <AddCourseForm addCourse={addCourse}/>]
        );
    }
}

CoursesPage.propTypes = {
    title: PropTypes.string.isRequired,
    courses:PropTypes.array.isRequired,
    addCourse: PropTypes.func.isRequired,
    removeCourse: PropTypes.func.isRequired
};

export default CoursesPage;
