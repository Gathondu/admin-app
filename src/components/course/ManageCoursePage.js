import React, { Component } from "react";
import { connect } from "react-redux";
import { Prompt } from "react-router-dom";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import toastr from "toastr";
import * as courseActions from "../../actions/course";
import CourseForm from "./CourseForm";

class ManageCoursePage extends Component {
  static propTypes = {
    course: PropTypes.object.isRequired,
    authors: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired
  };
  state = {
    course: Object.assign({}, this.props.course),
    loading: false,
    errors: { title: "", author: "", category: "", length: "" },
    formIsValid: false,
    titleValid: false,
    authorValid: false,
    categoryValid: false,
    lengthValid: false,
    formUpdated: false
  };
  componentWillReceiveProps(nextProps) {
    if (this.props.course.id !== nextProps.course.id) {
      this.setState({ course: Object.assign({}, nextProps.course) });
    }
  }
  shouldComponentUpdate(nextProps, nextState, nextContext) {
    if (this.state.course !== nextState.course) {
      this.setState({ formUpdated: true });
    }
    return true;
  }

  validateInput = course => {
    let errors = {};
    for (var field in course) {
      if (course.hasOwnProperty(field)) {
        switch (field) {
          case "title":
            if (course["title"].length > 0) {
              this.setState({ titleValid: true });
            } else {
              errors["title"] = "Title should contain more than 1 character";
            }
            break;
          case "authorId":
            if (course["authorId"] !== "") {
              this.setState({ authorValid: true });
            } else {
              errors["authorId"] = "Please select one of the authors";
            }
            break;
          case "category":
            if (course["category"].length > 0) {
              this.setState({ categoryValid: true });
            } else {
              errors["category"] =
                "Category should contain more than 1 character";
            }
            break;
          case "length":
            if (
              course["length"].length > 0 &&
              course["length"].match(/\[a-zA-z]+/g) === null &&
              course["length"].match(/^(\d{1,2}):(\d{1,2})/g) !== null
            ) {
              this.setState({ lengthValid: true });
            } else {
              errors["length"] =
                "Length should neither be empty nor contain words. Just numbers and a colon in the format 04:40";
            }
            break;
          default:
            break;
        }
      }
    }
    if (Object.keys(errors).length > 0) {
      this.setState({ errors: errors });
    } else {
      this.setState({ errors: {}, formIsValid: true });
    }
  };

  updateCoursesState = e => {
    const field = e.target.name;
    let course = Object.assign({}, this.state.course);
    course[field] = e.target.value;
    this.validateInput(course);
    this.setState({ course: course });
  };

  saveCourse = e => {
    e.preventDefault();
    this.setState({ loading: true });
    this.validateInput(this.state.course);
    if (this.state.formIsValid) {
      this.setState({ formUpdated: false });
      this.props.actions
        .saveCourse(this.state.course)
        .then(() => this.redirect())
        .catch(error => {
          this.setState({ loading: false });
          toastr.error(error);
        });
    }
  };

  redirect = () => {
    this.setState({ loading: false });
    toastr.success("Course Saved.");
    this.props.history.push("/courses");
  };

  render() {
    return [
      <Prompt
        key="prompt"
        when={this.state.formUpdated}
        message="The form isn't saved, are you sure you want to leave the page?"
      />,
      <CourseForm
        key="form"
        course={this.state.course}
        errors={this.state.errors}
        allAuthors={this.props.authors}
        onChange={this.updateCoursesState}
        onSave={this.saveCourse}
        loading={this.state.loading}
        formIsValid={this.state.formIsValid}
      />
    ];
  }
}

const getCourseById = (courses, id) => {
  const course = courses.filter(course => course.id === id);
  if (course.length) return course[0];
  return null;
};
const mapStateToProps = (state, ownProps) => {
  const courseId = ownProps.match.params.id; // get this from path `/course/:id`
  let course = {
    id: "",
    watchHref: "",
    title: "",
    authorId: "",
    length: "",
    category: ""
  };
  if (courseId && state.courses.length > 0) {
    course = getCourseById(state.courses, courseId);
  }
  const authorsFormattedForDropdown = state.authors.map(author => {
    return {
      value: author.id,
      text: author.firstName + " " + author.lastName
    };
  });

  return {
    course: course,
    authors: authorsFormattedForDropdown
  };
};
const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(courseActions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ManageCoursePage);
