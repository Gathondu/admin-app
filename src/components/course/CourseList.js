import React from 'react';
import PropTypes from 'prop-types';
import CourseListRow from "./CourseListRow";

const CourseList = ({courses, removeCourse}) => {
    const courseRows = courses.map((course, index) => (
        <CourseListRow
            course={course}
            key={course.id}
            index={index}
            removeCourse={removeCourse}
        />
    ));
    return (
        <table className="table">
            <thead>
                <tr>
                    <th>&nbsp;</th>
                    <th>Title</th>
                    <th>Author</th>
                    <th>Category</th>
                    <th>Length</th>
                </tr>
            </thead>
            <tbody>{courseRows}</tbody>
        </table>
    );
};

CourseList.propTypes = {
    courses: PropTypes.array.isRequired,
    removeCourse: PropTypes.func.isRequired
};

export default CourseList;
