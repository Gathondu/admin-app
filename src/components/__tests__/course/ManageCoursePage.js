import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {ManageCoursePage, mapDispatchToProps, getCourseById} from '../../course/ManageCoursePage';
import {courses} from "../../../api/mockCourse";

describe('Manage Course Page component', () => {
    const setup = () => {
        const props = {
            course: {
                id: '',
                title: '',
                watchHref: '',
                authorId: '',
                length: '',
                category: ''
            },
            authors: [],
            loading: false,
            errors: {},
            formIsValid: true,
            titleValid: true,
            authorValid: true,
            categoryValid: true,
            lengthValid: true,
            formUpdated: false,
            history: {
                push: jest.fn()
            }
        };

        return {
            props,
            wrapper:
                shallow(
                    <ManageCoursePage {...props}/>)
        };
    };
    it('should not regress', () => {
        const {wrapper} = setup();
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should validate user input correctly', () => {
        const {wrapper, props} = setup();
        const spy = jest.spyOn(wrapper.instance(), 'validateInput');
        wrapper.instance().validateInput(props.course);
        expect(spy).toHaveBeenCalled();
    });
    it('should update course state', () => {
        const {wrapper} = setup();
        const spy = jest.spyOn(wrapper.instance(), 'updateCoursesState');
        wrapper.instance().updateCoursesState({ target: {value: 'denis'}});
        expect(spy).toHaveBeenCalled();
    });
    it('should save course correctly', () => {
        const {wrapper} = setup();
        const spy = jest.spyOn(wrapper.instance(), 'saveCourse');
        wrapper.instance().saveCourse({preventDefault: () => {}});
        expect(spy).toHaveBeenCalled();
    });
    it('should redirect', () => {
        const {wrapper, props} = setup();
        wrapper.instance().redirect();
        expect(props.history.push).toHaveBeenCalled();
    });
    it('should get course by id', () => {
        const javascript = {
            id: "react-flux-building-applications",
            title: "Building Applications in React and Flux",
            watchHref: "http://www.pluralsight.com/courses/react-flux-building-applications",
            authorId: "cory-house",
            length: "5:08",
            category: "JavaScript"
        };
        expect(getCourseById(courses, 'react-flux-building-applications')).toEqual(javascript);
        expect(getCourseById(courses, 'rest')).toEqual(null);
    });
    it('should map dispatch to props', () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).actions.loadCoursesSuccess();
        expect(dispatch.mock.calls[0][0]).toEqual({"courses": undefined, "type": "LOAD_COURSES_SUCCESS"});
    });
});