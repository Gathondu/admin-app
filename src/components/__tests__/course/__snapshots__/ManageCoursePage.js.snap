// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Course Page component should not regress 1`] = `
<Fragment>
  <Prompt
    key="prompt"
    message="The form isn't saved, are you sure you want to leave the page?"
    when={false}
  />
  <form
    className="courseForm"
  >
    <h1>
      Manage Course
    </h1>
    <TextInput
      error=""
      label="Title"
      name="title"
      onChange={[Function]}
      placeholder="Enter Course Title"
      value=""
    />
    <SelectInput
      defaultOption="Select Author"
      label="Author"
      name="authorId"
      onChange={[Function]}
      options={Array []}
      value=""
    />
    <TextInput
      error=""
      label="Category"
      name="category"
      onChange={[Function]}
      placeholder="Enter Course Category"
      value=""
    />
    <TextInput
      error=""
      label="Length"
      name="length"
      onChange={[Function]}
      placeholder="Enter Course Length"
      value=""
    />
    <input
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="submit"
      value="Save"
    />
  </form>
</Fragment>
`;
